### PR TITLE
Pass channel classification to release pipeline

### PR DESCRIPTION
**Pull request by JohnTortugo**
Closes: https://github.com/dotnet/arcade/issues/2370

Passing this information to the release pipeline so that it'll be able to make decisions using it.

### CODE REVIEW

File: src/Microsoft.DotNet.Darc/src/DarcLib.AzDev/AzureDevOpsClient.cs [hunk 1]
```diff
@@ -931,9 +931,11 @@ public async Task<AzureDevOpsReleaseDefinition> RemoveAllArtifactSourcesAsync(st
         /// <param name="accountName">Azure DevOps account name</param>
         /// <param name="projectName">Project name</param>
         /// <param name="releaseDefinition">Release definition to be updated</param>
-        public async Task<int> StartNewReleaseAsync(string accountName, string projectName, AzureDevOpsReleaseDefinition releaseDefinition, int barBuildId)
+        /// <param name="barBuildId">BAR BuildID to be passed as variable to the pipeline</param>
+        /// <param name="channelClassification">BAR Channel Classification to be passed as variable to the pipeline</param>
+        public async Task<int> StartNewReleaseAsync(string accountName, string projectName, AzureDevOpsReleaseDefinition releaseDefinition, int barBuildId, string channelClassification = null)
         {
-            var body = $"{{ \"definitionId\": {releaseDefinition.Id}, \"variables\": {{ \"BARBuildId\": {{ \"value\": \"{barBuildId}\" }} }} }}";
+            var body = $"{{ \"definitionId\": {releaseDefinition.Id}, \"variables\": {{ \"BARBuildId\": {{ \"value\": \"{barBuildId}\" }}, \"ChannelClassification\": {{ \"value\": \"{channelClassification}\" }} }} }}";
```
Comment (jcagme) [inline src/Microsoft.DotNet.Darc/src/DarcLib.AzDev/AzureDevOpsClient.cs]: Maintaining a JSON string is not super escalable and could lead to mistakes. I'd suggest to create an object that then you serialize to string and use that instead. [This](https://github.com/dotnet/arcade-services/blob/4362345e236509143a7e3af380b3dc47c6b91f8e/src/Microsoft.DotNet.Darc/src/DarcLib/GitHubClient.cs#L123) method does that.

**Comment by mmitche** [comment]
@JohnTortugo What kind of info are we going to be inferring off of the classification?

Currently the classification is just a string without restrictions. Basing release pipeline decisions off of it introduces non-obvious magic.  If we do need this, then we should ensure that the classification has a bounded set of possible values with well documented meaning.

**Comment by JohnTortugo** [comment]
> @JohnTortugo What kind of info are we going to be inferring off of the classification?

The target feed to publish the packages to. 

This is particular interesting for arcade & arcade-validation where the same build (thus **same ArtifactsCategory**) is assigned to different channels. If we use the _same pipeline_ in the different channels then they will end up trying to publish again the same packages to the same feed. To workaround that, we decided to publish packages initially to a validation feed and then to dotnet-core feed. However, currently that is implemented by having two pipelines that only difference is where they publish packages to. Adding the `ChannelClassification` as a parameter to the pipeline is an attempt to 1) make the same pipeline be able to publish the same build to different feeds and 2) no need to duplicate pipelines that only difference is the target feed.

**Comment by JohnTortugo** [comment]
> Currently the classification is just a string without restrictions. Basing release pipeline decisions off of it introduces non-obvious magic. If we do need this, then we should ensure that the classification has a bounded set of possible values with well documented meaning.

I agree that that should be fixed :-)

**Comment by mmitche** [comment]
Okay I gotcha.  Could we parameterize the pipeline instead, and call it for the validation channel with one set of options, and the latest channel with another?

That would map closely onto what we should do for internal/public channels.  Both may be "dev" channels, but they will interpret artifact categories different.  That might be a better option than using the classification.

**Comment by JohnTortugo** [comment]
Talked with @mmitche offline and we decided on a better plan. I'll close the current PR and update the issue to reflect the new plan.

**Pull request closed without merging**